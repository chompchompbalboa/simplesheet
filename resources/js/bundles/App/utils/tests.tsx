//-----------------------------------------------------------------------------
// Imports
//-----------------------------------------------------------------------------
import React from 'react'
import { Provider as ReduxProvider } from 'react-redux'
import { applyMiddleware, createStore } from 'redux'
import thunkMiddleware from 'redux-thunk'
import { render } from '@testing-library/react'

import { appReducer } from '@app/state'

//-----------------------------------------------------------------------------
// With Redux
//-----------------------------------------------------------------------------
// @ts-ignore
const WithRedux = ({ children }) => {
  const store = createStore(appReducer, applyMiddleware(thunkMiddleware))
  return (
    <ReduxProvider store={store}>
      { children }
    </ReduxProvider>
  )
}

//-----------------------------------------------------------------------------
// Exports
//-----------------------------------------------------------------------------
export * from '@testing-library/react'
// @ts-ignore
export const renderWithRedux = (ui, options?) => render(ui, { wrapper: WithRedux, ...options })
