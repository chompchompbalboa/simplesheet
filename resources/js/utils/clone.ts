const clone = (itemToClone: any): any => {
	return JSON.parse(JSON.stringify(itemToClone))
}

export default clone
