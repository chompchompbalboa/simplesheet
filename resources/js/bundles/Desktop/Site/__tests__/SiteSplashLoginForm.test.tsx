//-----------------------------------------------------------------------------
// Imports
//-----------------------------------------------------------------------------
import React from 'react'
import 'jest-styled-components'
import '@testing-library/jest-dom/extend-expect'
import axiosMock from 'axios'

import { fireEvent, renderWithRedux } from '@/testing/library'
import { mockEnvironment } from '@/testing/mocks'
import { appStateFactory, IAppStateFactoryInput } from '@/testing/mocks/appState'
import { flushPromises } from '@/testing/utils'

import { 
  SiteSplashLoginForm,
  siteSplashLoginFormStatusMessages
} from '@desktop/Site/SiteSplashLoginForm'

//-----------------------------------------------------------------------------
// Setup
//-----------------------------------------------------------------------------
const {
  allFiles,
  allFileIds,
  allSheetsFromDatabase
} = appStateFactory({} as IAppStateFactoryInput)

const fileId = allFileIds[0]
const sheetId = allFiles[fileId].typeId
const sheetFromDatabase = allSheetsFromDatabase[sheetId]

// @ts-ignore
axiosMock.get.mockResolvedValue({ data: sheetFromDatabase })

console.warn = jest.fn()
console.error = jest.fn()

//-----------------------------------------------------------------------------
// Tests
//-----------------------------------------------------------------------------
describe('SiteSplashLoginForm', () => {

  const { location } = window

  beforeAll(() => {
    // @ts-ignore
    global.environment = mockEnvironment
    jest.useFakeTimers()
  })

  afterAll(() => {
    jest.useRealTimers()
  })

  beforeEach(() => {
    delete window.location
    // @ts-ignore
    window.location = { reload: jest.fn() }
  })

  afterEach(() => {
    window.location = location
    jest.clearAllMocks()
  })

  const siteSplashLoginForm = () => {
    const { 
      getByPlaceholderText, 
      getByTestId,
      getByText
    } = renderWithRedux(<SiteSplashLoginForm />)
    const emailInput = getByPlaceholderText('Email') as HTMLInputElement
    const passwordInput = getByPlaceholderText('Password') as HTMLInputElement
    const submitButton = getByText('Log In')
    const statusContainer = getByTestId('SiteSplashLoginFormStatus')
    return {
      emailInput,
      passwordInput,
      submitButton,
      statusContainer
    }
  }
  
  it("displays an 'Email' input", async () => {
    const { emailInput } = siteSplashLoginForm()
    expect(emailInput).toBeTruthy()
  })
  
  it("displays a 'Password' input", async () => {
    const { passwordInput } = siteSplashLoginForm()
    expect(passwordInput).toBeTruthy()
  })
  
  it("changes the 'Email' value on user input", async () => {
    const { emailInput } = siteSplashLoginForm()
    const nextEmailInputValue = 'test@test.com'
    expect(emailInput.value).toEqual('')
    fireEvent.change(emailInput, { target: { value: nextEmailInputValue }})
    expect(emailInput.value).toEqual(nextEmailInputValue)
  })
  
  it("changes the 'Password' value on user input", async () => {
    const { passwordInput } = siteSplashLoginForm()
    const nextPasswordInputValue = 'Super Secret Password'
    expect(passwordInput.value).toEqual('')
    fireEvent.change(passwordInput, { target: { value: nextPasswordInputValue }})
    expect(passwordInput.value).toEqual(nextPasswordInputValue)
  })

  it("changes the border color of the 'Email' input to 'red' if the user enters a value that is not an email and then blurs the input", async () => {
    const { emailInput } = siteSplashLoginForm()
    expect(emailInput.value).toBe('')
    expect(emailInput).toHaveStyleRule('border', '1px solid rgb(150,150,150)')
    fireEvent.change(emailInput, { target: { value: 'Not an email address' }})
    fireEvent.blur(emailInput)
    expect(emailInput).toHaveStyleRule('border', '1px solid red')
  })
  
  it("changes the value of the submit button to 'Logging In...' when the user submits the form", async () => {
    const { submitButton } = siteSplashLoginForm()
    submitButton.click()
    expect(submitButton.textContent).toBe('Logging In...')
  })
  
  it("displays an error message when the user tries to submit the form without all the fields completed", async () => {
    const { statusContainer, submitButton } = siteSplashLoginForm()
    submitButton.click()
    jest.advanceTimersByTime(500)
    expect(statusContainer.textContent).toBe(siteSplashLoginFormStatusMessages['NOT_ALL_FIELDS_ARE_COMPLETE'])
  })
  
  it("displays an error message when the user tries to submit the form without a valid email", async () => {
    const { emailInput, passwordInput, statusContainer, submitButton } = siteSplashLoginForm()
    fireEvent.change(emailInput, { target: { value: 'Not an email' }})
    fireEvent.change(passwordInput, { target: { value: 'Password' }})
    submitButton.click()
    jest.advanceTimersByTime(500)
    expect(statusContainer.textContent).toBe(siteSplashLoginFormStatusMessages['NOT_VALID_EMAIL'])
  })
  
  it("attempts to login the user when all fields are completed and the email is valid", async () => {
    const { emailInput, passwordInput, submitButton, statusContainer } = siteSplashLoginForm()
    const email = 'test@test.com'
    const password = 'Password'
    fireEvent.change(emailInput, { target: { value: email }})
    fireEvent.change(passwordInput, { target: { value: password }})
    submitButton.click()
    expect(statusContainer.textContent).toBe(siteSplashLoginFormStatusMessages['LOGGING_IN'])
    expect(axiosMock.post).toHaveBeenCalledTimes(1)
    expect(axiosMock.post).toHaveBeenCalledWith('/user/login', { email: email, password: password })
  })
  
  it("displays an error message when the login is not successful", async () => {
    (axiosMock.post as jest.Mock).mockRejectedValueOnce({ response: { status: 500 } })
    const { emailInput, passwordInput, submitButton, statusContainer } = siteSplashLoginForm()
    expect(window.location.reload).not.toHaveBeenCalled()
    const email = 'test@test.com'
    const password = 'Password'
    fireEvent.change(emailInput, { target: { value: email }})
    fireEvent.change(passwordInput, { target: { value: password }})
    submitButton.click()
    await flushPromises()
    expect(window.location.reload).not.toHaveBeenCalled()
    jest.advanceTimersByTime(500)
    expect(statusContainer.textContent).toBe(siteSplashLoginFormStatusMessages['ERROR_DURING_LOGIN'])
  })
  
  it("reloads the page when the login is successful", async () => {    
    (axiosMock.post as jest.Mock).mockResolvedValueOnce({})
    const { emailInput, passwordInput, submitButton, statusContainer } = siteSplashLoginForm()
    expect(window.location.reload).not.toHaveBeenCalled()
    const email = 'test@test.com'
    const password = 'Password'
    fireEvent.change(emailInput, { target: { value: email }})
    fireEvent.change(passwordInput, { target: { value: password }})
    submitButton.click()
    expect(statusContainer.textContent).toBe(siteSplashLoginFormStatusMessages['LOGGING_IN'])
    await flushPromises()
    expect(axiosMock.post).toHaveBeenCalledTimes(1)
    expect(axiosMock.post).toHaveBeenCalledWith('/user/login', { email: email, password: password })
    expect(window.location.reload).toHaveBeenCalled()
  })

})
