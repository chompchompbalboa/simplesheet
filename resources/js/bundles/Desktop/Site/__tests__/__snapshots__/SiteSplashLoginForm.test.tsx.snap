// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SiteSplashLoginForm renders without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c5 {
  width: 100%;
  margin-top: 0.5rem;
  margin-left: 0;
  cursor: pointer;
  padding: 0.5rem 1.25rem;
  border: 1px solid rgb(150,150,150);
  border-radius: 5px;
  font-size: 0.9rem;
  background-color: rgba(220,220,220,1);
  color: black;
  outline: none;
  -webkit-transition: background-color 0.1s;
  transition: background-color 0.1s;
}

.c5:hover {
  background-color: white;
  color: black;
}

.c2 {
  margin-left: 0;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c3 {
  width: 100%;
}

.c4 {
  width: 100%;
  margin: 0.375rem;
  padding: 0.5rem 0.25rem;
  border: none;
  border: 1px solid rgb(150,150,150);
  border-radius: 4px;
  outline: none;
  font-size: 0.9rem;
}

.c7 {
  margin-top: 0.5rem;
  text-align: left;
  color: rgb(175,0,0);
}

.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 {
  width: 100%;
}

@media (max-width:480px) {
  .c5 {
    width: 100%;
    margin: 0.375rem;
  }
}

@media (max-width:480px) {
  .c4 {
    width: 100%;
    margin: 0.375rem 0;
  }
}

@media (max-width:480px) {
  .c1 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

<body>
    <div>
      <form
        class="c0"
      >
        <div
          class="c1"
        >
          <div
            class="c2"
          >
            <label
              class="c3"
            >
              Email
              :
            </label>
            <input
              class="c4"
              placeholder="Email"
              type="email"
              value=""
            />
          </div>
          <div
            class="c2"
          >
            <label
              class="c3"
            >
              Password
              :
            </label>
            <input
              class="c4"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <button
            class="c5"
          >
            Log In
          </button>
        </div>
        <div
          class="c6"
        >
          <div
            class="c7"
            data-testid="SiteSplashLoginFormStatus"
          />
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="sc-AykKH iRHvOU"
    >
      <div
        class="sc-AykKI buDOwc"
      >
        <div
          class="sc-AykKD bbzThp"
        >
          <label
            class="sc-AykKE jtbLvj"
          >
            Email
            :
          </label>
          <input
            class="sc-AykKF haJOPm"
            placeholder="Email"
            type="email"
            value=""
          />
        </div>
        <div
          class="sc-AykKD bbzThp"
        >
          <label
            class="sc-AykKE jtbLvj"
          >
            Password
            :
          </label>
          <input
            class="sc-AykKF haJOPm"
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <button
          class="sc-AykKC ilFvXC"
        >
          Log In
        </button>
      </div>
      <div
        class="sc-AykKJ OwdgA"
      >
        <div
          class="sc-AykKG himGUV"
          data-testid="SiteSplashLoginFormStatus"
        />
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
    Symbol(observable): [Function],
  },
  "unmount": [Function],
}
`;
